kratos/containers/pointer_vector_set.h

```diff
@@ -1066,51 +1066,53 @@ class PointerVectorSet final
+
+            while (existing_data_it != mData.end() && new_data_it != last) {
+                const auto existing_key = KeyOf(GetReference(existing_data_it));
+                const auto new_key = KeyOf(GetReference(new_data_it));
+                if (existing_key == new_key) {
+                    // adds the existing item and ignore the new item to keep consistent with the std::set::insert
+                    aux_container.push_back(TPointerType(&GetReference(existing_data_it)));
+                    ++existing_data_it;
+                    ++new_data_it;
+                } else if (existing_key < new_key) {
+                    // existing key comes before the new key
+                    aux_container.push_back(TPointerType(&GetReference(existing_data_it)));
+                    ++existing_data_it;
                 } else {
-                    // now if the capacity of the new mData is larger than the existing
-                    // capacity, then the current lower_bound_first is invalidated.
-                    // hence needs to find it again.
-                    const auto new_lower_bound = std::lower_bound(mData.begin(), mData.end(), KeyOf(GetReference(first)), CompareKey());
-                    auto current_pos = new_lower_bound - 1;
-                    for (auto it = first; it != last; ++it) {
-                        current_pos = mData.insert(current_pos + 1, TPointerType(&GetReference(it)));
-                    }
-                }
-            } else {
-                auto p_current_itr = mData.begin();
-                // now add the new elements
-                for (auto it = first; it != last; ++it) {
-                    // find the lower bound element.
-                    p_current_itr = std::lower_bound(p_current_itr, mData.end(), KeyOf(GetReference(it)), CompareKey());
-                    if (p_current_itr == mData.end() || !EqualKeyTo(KeyOf(GetReference(it)))(*p_current_itr)) {
-                        p_current_itr = mData.insert(p_current_itr, TPointerType(&GetReference(it)));
-                    }
+                    // existing key comes after the new key
+                    aux_container.push_back(TPointerType(&GetReference(new_data_it)));
+                    ++new_data_it;
                 }
             }
+
+            // now add the left overs. At this moment, either both existing_data_it and new_data_it
+            // reached the end, or one of them reached the end. In either case, we can safely add the remaining items
+            // using the push back.
+            for (; new_data_it != last; ++new_data_it) {
+                aux_container.push_back(TPointerType(&GetReference(new_data_it)));
+            }
+
+            for (; existing_data_it != mData.end(); ++existing_data_it) {
+                aux_container.push_back(TPointerType(&GetReference(existing_data_it)));
+            }
+
+            // now do the swap
+            mData.swap(aux_container);
+            mData.shrink_to_fit();
```

Comment:
i am not 100% we want this shrink_to_fit (it will trigger a reallocation ), other than that ok to me

Reply:
True, it not guranteed to have re-allocation or not. I will remove it.

Reply:
done :)